### PR TITLE
Filter: Only remove multiple spaces

### PR DESCRIPTION
**Pull request by djbe**
Modify the `removeNewlines` filter to only remove whitespace characters:
- if they're a newline
- if there are multiple consecutive whitespace chars
- if they're at either end of the string (trimming)

Remaining single whitespace characters are converted to a space ` `.

### CODE REVIEW

File: Documentation/filters-strings.md [hunk 1]
```diff
@@ -57,15 +57,17 @@ Removes all newlines and whitespace characters from the string.
...
-By default it removes whitespace characters, unless a single optional argument is set to "false", "no" or "0":
+By default it removes consecutive whitespace characters (and trims the ends), unless a single optional argument is set to "false", "no" or "0":
 
 | Input                 | Output                |
 |-----------------------|-----------------------|
 | `  \ntest`            | `  test`              |
 | `test \n\t `          | `test \t `            |
 | `test\n  test`        | `test  test`          |
 | `\r\ntest\n  test\n`  | `test  test`          |
+| ` test test `         | ` test test `         |
```
Comment (AliSoftware) [inline Documentation/filters-strings.md]: Not sure why the beginning & end whitespaces would be kept there? I'd expect them to be removed by `.trimmingCharacters(in: .whitespacesAndNewlines)` here.

File: Tests/StencilSwiftKitTests/StringFiltersTests.swift [hunk 2]
```diff
@@ -140,11 +140,12 @@ extension StringFiltersTests {
...
-      "  \n test": "test",
-      "test  \n ": "test",
-      "test\n \ntest": "testtest",
-      "\n test\n \ntest \n ": "testtest"
+      "test1": "test1",
+      "  \n test2": "test2",
+      "test3  \n ": "test3",
+      "test4\n \ntest": "test4test",
+      "\n test5\n \ntest test \n ": "test5test test",
+      "test6\ntest": "test6test"
```
Comment (AliSoftware) [inline Tests/StencilSwiftKitTests/StringFiltersTests.swift]: Maybe add that `" test test ":"test test"` here to confirm my suggestion about that it should remove the leading/trailing whitespaces in the examples you put in the md docs above

File: Documentation/filters-strings.md (hunk 1 above)
Comment (djbe) [inline Documentation/filters-strings.md]: This is for the false case, where only newlines are removed.This is

Comment (AliSoftware) [inline Documentation/filters-strings.md]: Oh, right.

Comment (AliSoftware) [inline Documentation/filters-strings.md]: Oh, right.

**Comment by djbe** [comment]
Hmmm, instead of doing multiple regexes, would it be easier (for the removewhitespaces case) to:
- split based on newlines
- trim each line
- join again

File: Tests/StencilSwiftKitTests/StringFiltersTests.swift (hunk 2 above)
Comment (AliSoftware) [inline Tests/StencilSwiftKitTests/StringFiltersTests.swift]: Still, I like that the examples we give in the MD documentation are used in the tests as well, if only to be sure that we are consistent with what we state in the doc and that those cases are indeed checked

Comment (djbe) [inline Tests/StencilSwiftKitTests/StringFiltersTests.swift]: I'll add that specific case, but I'll number it (easier for debugging if needed) like the rest.

**Comment by AliSoftware** [comment]
As long as the test cases still passes when you change the implementation… 😉 
I'm not sure what would be faster, but RegExes even if powerful are also generally slower indeed, so might be worth it yeah

**Comment by djbe** [comment]
I'm just wondering what would be useful for users (and our use cases).

The main one will be the ability to write things like macros on multiple lines but still have the result be one line. This might mean multiple things though:
- I don't want any spaces at all, anywhere
- Just remove the newlines and leading whitespaces
- ...?

**Comment by AliSoftware** [comment]
That's pretty much the use cases I see indeed. Creating some identifier or word from a complex expression (= no spaces at all anywhere) or formatting the template nicely (remove newlines and leading spaces per lines), so I think this covers all we need for now?

**Comment by djbe** [comment]
Hmmm no then I need to change the code a bit. Right now it either:
- only removes newlines and nothing else
- removes newlines and multiple consecutive spaces + replaces single whitespaces with a space

**Comment by djbe** [comment]
Instead of a boolean parameter, should we maybe accept a string to represent a bunch of cases ("all", "leading", ...)

**Comment by AliSoftware** [comment]
Mmmh I was gonna say YAGNI, but given that the different cases are not that straightforward to explain ("remove whitespaces except this but that…"), a string instead of a bool might be more explicit anyway, so yeah, go for it.

File: Sources/Filters+Strings.swift [hunk 3]
```diff
@@ -120,13 +120,23 @@ extension Filters {
     }
 
     static func removeNewlines(_ value: Any?, arguments: [Any?]) throws -> Any? {
-      let removeSpaces = try Filters.parseBool(from: arguments, index: 0, required: false) ?? true
+      let mode = arguments.first as? String ?? "all"
```
Comment (AliSoftware) [inline Sources/Filters+Strings.swift]: Shouldn't we throw if the argument is not a valid string, like `removeNewlines:42`?

So if there's no argument at all (`arguments.first == nil`), agreed that `all` should be the default. But if the argument isn't castable as a `as? String`, maybe throw?

File: Sources/Filters+Strings.swift [hunk 4]
```diff
@@ -120,13 +120,23 @@ extension Filters {
...
       guard let string = value as? String else { throw Filters.Error.invalidInputType }
 
-      let set: CharacterSet = removeSpaces ? .whitespacesAndNewlines : .newlines
-      let result = string.components(separatedBy: set).joined()
-
-      return result
+      switch mode {
+      case "all":
```
Comment (AliSoftware) [inline Sources/Filters+Strings.swift]: Use some constants instead of magic strings here? Even if those constants are scoped inside the func.

e.g.
```swift
static func removeNewlines(_ value: Any? arguments: [Any?]) throws -> Any? {
  enum Modes: String {
    case all, leading
  }

  let modeArg = String(describing: arguments.first) ?? Modes.all.rawValue
  guard let mode = Modes(rawValue: modeArg) else {
    throw Filters.Error.invalidOption(option: modeArg)
  }

  switch mode {
  case .all: …
  case .leading: …
  }
}
```

**Comment by djbe** [comment]
Before you ask, the reason that enum is not inside the function is because `swiftlint` complains that such things can be at most 1 level deep (nesting) 🤷‍♂️

File: Sources/Filters+Strings.swift [hunk 5]
```diff
@@ -119,14 +123,31 @@ extension Filters {
...
-      return result
+      switch mode {
+      case .all:
+        return string
+          .components(separatedBy: .whitespacesAndNewlines)
+          .joined()
+      case .leading:
+        return string
+          .components(separatedBy: .newlines)
+          .map { String($0.unicodeScalars.drop(while: { CharacterSet.whitespaces.contains($0) })) }
```
Comment (AliSoftware) [inline Sources/Filters+Strings.swift]: Codebeat seems to be unhappy with nesting too deep here, maybe separate the closure used in that map in a private function to make it happy?

File: Tests/StencilSwiftKitTests/ParseEnumTests.swift [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    let value = try Filters.parseEnum(from: ["foo"], default: Test.baz)
+    XCTAssertEqual(value, Test.foo)
+  }
+
+  func testParseEnum_WithBarString() throws {
+    let value = try Filters.parseEnum(from: ["bar"], default: Test.baz)
+    XCTAssertEqual(value, Test.bar)
+  }
+
+  func testParseEnum_WithBazString() throws {
```
Comment (AliSoftware) [inline Tests/StencilSwiftKitTests/ParseEnumTests.swift]: Add a test with a non-zero index argument too
